File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderIn.java
package uk.gov.justice.hmpps.prison.api.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Builder;
import lombok.Data;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

import static io.swagger.v3.oas.annotations.media.Schema.RequiredMode.REQUIRED;

@Schema(description = "Summary of an offender counted as Establishment Roll - In")
@JsonInclude(JsonInclude.Include.NON_NULL)
@Builder(toBuilder = true)
@Data

public class OffenderIn {
    @NotBlank
    @Schema(requiredMode = REQUIRED, description = "Display Prisoner Number")
    private String offenderNo;

    @NotNull
    private Long bookingId;

    @NotNull
    private LocalDate dateOfBirth;

    @NotBlank
    private String firstName;

    private String middleName;

    @NotBlank
    private String lastName;

    @NotBlank
    @Schema(requiredMode = REQUIRED, description = "Id for Agency travelling from")
    private String fromAgencyId;

    @NotBlank
    @Schema(requiredMode = REQUIRED, description = "Description for Agency travelling from")
    private String fromAgencyDescription;

    @NotBlank
    @Schema(requiredMode = REQUIRED, description = "Id for Agency travelling to")
    private String toAgencyId;

    @NotBlank
    @Schema(requiredMode = REQUIRED, description = "Description for Agency travelling to")
    private String toAgencyDescription;

    @NotBlank
    @Schema(description = "City offender was received from")
    private String fromCity;

    @NotBlank
    @Schema(description = "City offender was sent to")
    private String toCity;

    @NotNull
    @Schema(requiredMode = REQUIRED, description = "Movement time")
    private LocalTime movementTime;

    @NotNull
    @Schema(requiredMode = REQUIRED, description = "Movement date time")
    private LocalDateTime movementDateTime;

    @NotNull
    @Schema(requiredMode = REQUIRED, description = "Description of the offender's (internal) location")
    private String location;

    public OffenderIn(@NotBlank String offenderNo, @NotNull Long bookingId, @NotNull LocalDate dateOfBirth, @NotBlank String firstName, String middleName, @NotBlank String lastName, @NotBlank String fromAgencyId, @NotBlank String fromAgencyDescription, @NotBlank String toAgencyId, @NotBlank String toAgencyDescription, @NotBlank String fromCity, @NotBlank String toCity, @NotNull LocalTime movementTime, @NotNull LocalDateTime movementDateTime, @NotNull String location) {
        this.offenderNo = offenderNo;
        this.bookingId = bookingId;
        this.dateOfBirth = dateOfBirth;
        this.firstName = firstName;
        this.middleName = middleName;
        this.lastName = lastName;
        this.fromAgencyId = fromAgencyId;
        this.fromAgencyDescription = fromAgencyDescription;
        this.toAgencyId = toAgencyId;
        this.toAgencyDescription = toAgencyDescription;
        this.fromCity = fromCity;
        this.toCity = toCity;
        this.movementTime = movementTime;
        this.movementDateTime = movementDateTime;
        this.location = location;
    }

    public OffenderIn() {}
}

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderIn.kt
package uk.gov.justice.hmpps.prison.api.model

import com.fasterxml.jackson.annotation.JsonInclude
import io.swagger.v3.oas.annotations.media.Schema
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

@Schema(description = "Summary of an offender counted as Establishment Roll - In")
@JsonInclude(JsonInclude.Include.NON_NULL)
data class OffenderIn(
  @Schema(description = "Display Prisoner Number", required = true)
  val offenderNo: String? = null,

  @Schema(description = "Booking ID", required = true)
  val bookingId: Long? = null,

  @Schema(description = "Date of birth", required = true)
  val dateOfBirth: LocalDate? = null,

  @Schema(description = "First name", required = true)
  val firstName: String? = null,

  @Schema(description = "Middle name", required = false)
  val middleName: String? = null,

  @Schema(description = "Display Prisoner Number", required = true)
  val lastName: String,

  @Schema(description = "Id for Agency travelling from", required = false)
  val fromAgencyId: String? = null,

  @Schema(description = "Description for Agency travelling from", required = false)
  val fromAgencyDescription: String? = null,

  @Schema(description = "Id for Agency travelling to", required = false)
  val toAgencyId: String? = null,

  @Schema(description = "Description for Agency travelling to", required = false)
  val toAgencyDescription: String? = null,

  @Schema(description = "From city", required = false)
  val fromCity: String? = null,

  @Schema(description = "City offender was sent to", required = false)
  val toCity: String? = null,

  @Schema(description = "Movement time", required = true)
  val movementTime: LocalTime,

  @Schema(description = "Movement date time", required = true, example = "2021-07-16T12:34:56")
  val movementDateTime: LocalDateTime? = null,

  @Schema(description = "Description of the offender's (internal) location", required = false)
  val location: String? = null,

  @Schema(description = "Type of movement", example = "TAP", required = true)
  val movementType: String,

  @Schema(description = "Reason for movement in", required = false)
  val movementReasonDescription: String? = null,

  @Schema(description = "From address", required = false)
  val fromAddress: String? = null,
)
